src/proxy/http/HttpSM.cc

```diff
@@ -2980,6 +2980,12 @@ HttpSM::tunnel_handler(int event, void *data)
 {
   STATE_ENTER(&HttpSM::tunnel_handler, event);
 
+  // If we had already received EOS, just go away. We would sometimes see
+  // a WRITE event appear after receiving EOS from the server connection
+  if ((event == VC_EVENT_WRITE_READY || event == VC_EVENT_WRITE_COMPLETE) && server_entry->eos) {
```

Comment:
As your pointed out, other event handlers have a guard like this - e.g. `HttpSM::state_read_server_response_header`. 

https://github.com/apache/trafficserver/blob/14d5b2ce401446a13b2fef4ab8dbc7cf890acf25/src/proxy/http/HttpSM.cc#L1873-L1877

So this should be good as a workaround fix. However, I prefer to move the state that ignore events when HttpSM got EOS. Because HttpSM is a state machine. It should not controlled by a bunch of flags .